Ruby group capstone: day 2

khanh-devos opened the pull request:
### [Logic](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/catalog_of_my_things.md#logic)
- [x] Each class from the [UML class diagram](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/images/catalog_of_my_things.png) should be created in a separate .rb file.
- [x] All associations (1-to-many relationships) and aggregations (parent-child relationships) between classes should be implemented.
- [x] All properties visible in the diagram should be defined and set up in the constructor method. Exception: properties for the 1-to-many relationships should NOT be set in the constructor method. Instead, they should have a custom setter method created.
- [x] All methods visible in the diagram should be implemented:
add_item method (in all classes that have that method)
should take an instance of the Item class as an input.
should add the input item to the collection of items.
should add self as a property of the item object (by using the correct setter from the item object).
- [x] can_be_archived?() in the Item class
should return true if published_date is older than 10 years.
otherwise, it should return false.
- [x] move_to_archive() in the Item class
should reuse can_be_archived?() method.
should change the archived property to true if the result of the can_be_archived?() method is true.
should do nothing if the result of the can_be_archived?() method is false.
- [x] can_be_archived?() in the Book class
should override the method from the parent class.
should return true if parent's method returns true OR if cover_state equals to "bad".
otherwise, it should return false.
- [x] can_be_archived?() in the MusicAlbum class
should override the method from the parent class.
should return true if parent's method returns true AND if on_spotify equals true.
otherwise, it should return false.
can_be_archived?() in the Movie class
- [x] should override the method from the parent class.
should return true if parent's method returns true OR if silent equals true.
otherwise, it should return false.
- [x] can_be_archived?() in the Game class
should override the method from the parent class.
should return true if parent's method returns true AND if last_played_at is older than 2 years.
otherwise, it should return false.
Add unit tests for all implemented methods.
### [User interface](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/catalog_of_my_things.md#user-interface)
Create a main.rb file that will serve as your console app entry-point.
Your console app, at the start, should:
Present the user with a list of options to perform.
Let users choose an option.
If needed, ask for parameters for the option.
Have a way to quit the app.
The following options should be available:
- [x] List all books
- [x] List all music albums
- [x] List all movies
- [x] List of games
- [x] List all genres (e.g 'Comedy', 'Thriller')
- [x] List all labels (e.g. 'Gift', 'New')
- [x] List all authors (e.g. 'Stephen King')
- [x] List all sources (e.g. 'From a friend', 'Online shop')
- [x] Add a book
- [x] Add a music album
- [x] Add a movie
- [x] Add a game
- [x] All data should be preserved by saving collections in .json files.
### [Database](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/catalog_of_my_things.md#database)
Create a schema.sql file with tables that will be analogical to the structure of the classes in your app. As you cannot implement inheritance in the database tables - add all properties and associations from the parent Item class as table columns to all tables based on the child classes. Thanks to that it can be used to store data in the future.
### [Project documentation](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/catalog_of_my_things.md#project-documentation)
Once you have finished the development of the project, you should record a video presenting the features of the project you built. It is a video with a maximum length of 5 minutes. The content of the video should include:

### A description of the project.
A demo of the project features.
You should also highlight some interesting piece of code or something you built that you are very proud of.
You all should appear in the video and talk about the project.
For recording the video you can use the Zoom recording features while in a call with your peers.

khanh-devos committed:
Added rubocop STYLE linter and README

khanh-devos committed:
add methods into Item

khanh-devos committed:
modify some attrs in Item

khanh-devos committed:
add list of items will be shown

khanh-devos committed:
Added classes BookList, LabelList, InputValidation
- BookList for creating a list of books.
- LabelList for creating a list of labels and add|updating a label of a book, game, or album.
- Built the main.rb for execution.

khanh-devos committed:
Added serializer JSON item and array.
- Serializeitem for all the classes
- Serializearray for all the lists of classes

khanh-devos committed:
Added schema file

khanh-devos committed:
modified schema music_album

khanh-devos committed:
Testing classes: book, label and item

khanh-devos committed:
fix some rubocop style errors

khanh-devos committed:
Fixed some rubocop style errors

khanh-devos committed:
Merge pull request #7 from khanh-devos/old_way

Ruby group capstone: day 1

kit0-0 committed:
Add author.rb

kit0-0 committed:
Add game.rb

kit0-0 committed:
Create Authorlist class

kit0-0 committed:
Create GameList class

kit0-0 committed:
Added support for authors and games

kit0-0 committed:
Add json to store author

kit0-0 committed:
Add json to store game

kit0-0 committed:
Add test for author

kit0-0 committed:
Add test for game

kit0-0 committed:
Add comments to fix empty file linters

kit0-0 committed:
Fix linters for EmptyLines

Aqurds committed:
Add musicAlbum, Genre class

kit0-0 committed:
Updatd select_create_author_for

kit0-0 committed:
Add add_authorto book

kit0-0 committed:
Add Game Title

khanh-devos committed:
Merge pull request #9 from khanh-devos/feature/music

Add musicAlbum, Genre class

kit0-0 committed:
Merge branch 'dev' into game-author-association

kit0-0 committed:
Moved add_author to item class

kit0-0 committed:
Add error handlinf of a=json empty

kit0-0 committed:
Fix linters

kit0-0 committed:
Updated test for author and game

khanh-devos committed:
Added genre, genrelist, musicalbum, albumlist

kit0-0 committed:
Merge pull request #8 from khanh-devos/game-author-association

Implement Game and Author Classes with Associations

khanh-devos committed:
Tested musicalbum and genre

khanh-devos committed:
Merge branch 'dev' into album_genre

khanh-devos committed:
Merge pull request #10 from khanh-devos/album_genre

Ruby group capstone: day 2

khanh-devos committed:
Features of this console ruby app:
- Created 6 classes: book, label, game, author, music album, genre. (UML included)
- Allow: add a new book|game|music album, show lists of all of them.
- Save all data by JSON format and re-load all the data for a new playing.

khanh-devos committed:
Added video presentation

AbelG101 commented on README.md:
```diff
@@ -1 +1,160 @@
+## 🚀 Live Demo <a name="live-demo"></a>
+[Video presentation](https://drive.google.com/file/d/1Km66n3_KDSNNz56if8WGyZWY2Fyqgipx/view?usp=drive_link)
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
```
- [x] Really great presentation guys, but as part of the project requirements the maximum length has to be 5mins so please work on this. 💪

Aqurds committed:
Update README title & author

AbelG101 commented on classes/item.rb:
```diff
@@ -0,0 +1,65 @@
+  include SerializationItem
+  attr_reader :id, :type, :archived
+  attr_accessor :genre, :author, :label, :published_date
+
```
- [x] Since the genre, author, and label instance variables are set using the methods you defined like `add_genre`, `add_author`, and so forth I'd advise you set the `attr_reader` for them here.

AbelG101 commented on classes/item.rb:
```diff
@@ -0,0 +1,65 @@
+
+  def add_author(author)
+    author.add_item(self)
+  end
```
- [x] Kindly make sure to assign the instance variables with the passed arguments in the functions above as shown below
```rb
  def add_genre(genre)
   @genre = genre
    genre.add_item(self)
  end
```
- This applies for the rest of the methods as well.

AbelG101 commented on classes/game.rb:
```diff
@@ -0,0 +1,27 @@
+
+  private
+
+  def can_be_archived?
+    unless last_played_at.nil?
+      age_in_seconds = Time.now - last_played_at
+
+      return age_in_seconds > (2 * 365 * 24 * 60 * 60)
+    end
+
+    false
+  end
```
- [x] Ouch! this is a close one but the `can_be_archived` method in the `Game` class should return true if the parent's method  returns true so kindly make sure to include `super &&` before returning in line `22`.

AbelG101 commented on schemas.sql:
```diff
@@ -0,0 +1,66 @@
+    PRIMARY KEY (id, title)
+)
+
+CREATE TABLE book (
+    id SERIAL PRIMARY KEY,
+    item_type VARCHAR(250),
+    title VARCHAR(250),
+    cover_state BOOLEAN,
+
+    published_date DATE,
+    archived BOOLEAN,
+    
+    genre_id INT REFERENCES genre(id),
+    author_id INT REFERENCES author(id),
+    label_id INT REFERENCES label(id)
+)
```
- [x] It seems you're missing one property `publisher` in the book class here so please make sure to update this.

AbelG101 commented on schemas.sql:
- [x] As part of the project requirements please make sure all foreign key columns have corresponding indexes in the schema file here.

AbelG101 reviewed (changes requested):
![](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExNzFmY2U0ZTkwMmM3ZTM5MDBkNjhhZjk0MjI0OTU3MGNjYjMyMmRhOCZjdD1n/M1fddM48g9084bJMEX/giphy.gif)


# Hi @khanh-devos 👋, 

You're doing a good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- App works fabulously ✔
- All classes are properly implemented ✔
- Properly written unit tests ✔
- Correct use of Git Flow ✔
- Well-structured files ✔
- All linter checks are passing ✔
- Descriptive PR title with a good summary ✔

## Required Changes ♻️
- _Check the comments under the review._
- you just have to make a few changes and you're on your way 🚀.


## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

khanh-devos committed:
Fixed the app following the code review requiremnt:
- Reduced the video length to 5 mins.
- Item class: add genre, label, author to attr_reader.
- Item class: modify add_label, add_author, add_genre.
- Game class: add parent'method into method can_be_archived?
- Schemas: added publisher into the table books.
- Schemas: all foreign keys have corresponding indexes.

khanh-devos committed:
Merge branch 'dev' of https://github.com/khanh-devos/ruby-cap-catalog into dev

elisha2kyakpo1 reviewed (approved):
Hi 👋 team, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights

👋 You did a great job fixing the changes requested by the previous reviewer.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

khanh-devos commented:
Thank you for your quick review, @elisha2kyakpo1 ,
